File: lab5/java/comp26120/enum_kp.java
package comp26120;

import java.util.ArrayList;

public class enum_kp extends KnapSack {
    public enum_kp(String filename) {
	super(filename);
    }

    
    public static void main(String[] args) {
	enum_kp knapsack = new enum_kp(args[0]);

	knapsack.print_instance();
	knapsack.enumerate();
    }

    public void enumerate() {
	// Do an exhaustive search (aka enumeration) of all possible ways to pack
	// the knapsack.
	// This is achieved by creating every binary solution vector of length Nitems.
	// For each solution vector, its value and weight is calculated.

	int i; // item index
	ArrayList<Boolean> solution = new ArrayList<Boolean>(Nitems + 1); // Solution vector representing items packed
	solution.add(null); // C implementation has null first object
	    
	ArrayList<Boolean> best_solution = new ArrayList<Boolean>(Nitems + 1); // Solution vector for best solution found
	best_solution.add(null);

	int best_value; // total value packed in the best solution
	double j = 0;
	int infeasible; // 0 means feasible; -1 means infeasible (violates the capacity constraint)

	// set the knapsack initially empty
	for (i = 1; i <= Nitems; i++) {
	    solution.add(false);
	    best_solution.add(false);
	}

	QUIET=true;
	best_value = 0;

	while (!(next_binary(solution, Nitems))) {
	    /* ADD CODE IN HERE TO KEEP TRACK OF FRACTION OF ENUMERATION DONE */

	    // calculates the value and weight and feasibility:
	    infeasible=check_evaluate_and_print_sol(solution);  
	    /* ADD CODE IN HERE TO KEEP TRACK OF BEST SOLUTION FOUND*/
	}
	/* ADD CODE TO PRINT OUT BEST SOLUTION */
    }

    public boolean next_binary(ArrayList<Boolean> str, int Nitems) {
	// Called with an ArrayList of 0/1 entires with length Nitems
	// (0th item is null for some reason inherited from C implementation - I expect
	// because in C the array is here treated as a pointer to a string).
	// If we treat this array list as a binary number, then this
	// Function adds "1" - e.g., {0,0,0,1} would turn to {0,0,1,0}
	// It returns true when there are no possible binary numbers left 
	int i = Nitems;
	while(i>=1) {
	    if (str.get(i) == true) {
		str.set(i, false);
		i--;
	    } else {
		str.set(i, true);
		break;
	    }
	}
	if (i == 0) {
	    return true;
	} else {
	    return false;
	}
    }
}

File: lab5/python/enum_kp.py
import sys
import math

from knapsack import knapsack


class enum_knapsack(knapsack):
    def __init__(self, filename):
        knapsack.__init__(self, filename)

    def enumerate(self):
        # Do an exhaustive search (aka enumeration) of all possible ways to pack
        # the knapsack.
        # This is achived by creating every "binary" solution vectore of length Nitems.
        # For each solution vector, its value and weight is calculated

        solution = [False] * (self.Nitems + 1)  # (binary/ true/false) solution vectore representing items pack
        best_solution = [False] * (self.Nitems + 1)  # (binary) solution veectore for best solution found
        j = 0.0

        self.QUIET = True
        best_value = 0  # total value packed in the best solution
        current_progress = 0.0

        while not self.next_binary(solution, self.Nitems):
            current_progress_int = int(current_progress * 100)
            j = j + 1.0
            current_progress = j / math.pow(2, self.Nitems)
            if not self.QUIET:
                print("done %g of the full enumeration" % current_progress)
            if current_progress_int!=int(current_progress * 100):
                print("\r", end="")
                print(
                    f"program running status: {int(current_progress * 100)}% | [{'▋' * (int(current_progress * 100) // 2)}{' ' * (50-(int(current_progress * 100) // 2))}]",
                    end="")

            infeasible = self.check_evaluate_and_print_sol(solution)

            if (not infeasible) and self.total_value > best_value:
                best_value = self.total_value

                for i in range(1, self.Nitems + 1):
                    best_solution[i] = solution[i]

            if not self.QUIET:
                print("best so far has value %d" % best_value)

        self.QUIET = False
        print("Finished.\nPack the following items for optimal")
        self.check_evaluate_and_print_sol(best_solution)

    def next_binary(self, sol, Nitems):
        # Called with a "binary" vector of length Nitems, this
        # method "adds 1" to the vector, e.g. 0001 would turn to 0010.
        # If the string overflows, then the function returs True, else it returns False
        i = Nitems
        while i > 0:
            if sol[i]:
                sol[i] = False
                i = i - 1
            else:
                sol[i] = True
                break
        if i == 0:
            return True
        else:
            return False


knapsk = enum_knapsack(sys.argv[1])
knapsk.print_instance()
knapsk.enumerate()
